File: web/src/lib/serbianPronunciation.ts
const single: Record<string, string> = {
  a: 'a', b: 'b', c: 'ts', č: 'tʃ', ć: 'tɕ', d: 'd', đ: 'dʑ',
  e: 'e', f: 'f', g: 'ɡ', h: 'x', i: 'i', j: 'j', k: 'k', l: 'l',
  m: 'm', n: 'n', o: 'o', p: 'p', r: 'r', s: 's', š: 'ʃ', t: 't',
  u: 'u', v: 'ʋ', z: 'z', ž: 'ʒ',
  а: 'a', б: 'b', в: 'ʋ', г: 'ɡ', д: 'd', ђ: 'dʑ', е: 'e', ж: 'ʒ',
  з: 'z', и: 'i', ј: 'j', к: 'k', л: 'l', љ: 'ʎ', м: 'm', н: 'n',
  њ: 'ɲ', о: 'o', п: 'p', р: 'r', с: 's', т: 't', ћ: 'tɕ', у: 'u',
  ф: 'f', х: 'x', ц: 'ts', ч: 'tʃ', џ: 'dʒ', ш: 'ʃ',
};

const pairs: Record<string, string> = { dž: 'dʒ', lj: 'ʎ', nj: 'ɲ' };

const IPA_VOWELS = 'aeiou';

/** Звук транскрипции и признак «это вершина слога». */
interface Sound {
  text: string;
  nucleus: boolean;
}

function sounds(word: string): Sound[] {
  const value = word.trim().toLocaleLowerCase('sr');
  const out: Sound[] = [];
  for (let index = 0; index < value.length; index += 1) {
    const pair = value.slice(index, index + 2);
    if (pairs[pair]) {
      out.push({ text: pairs[pair]!, nucleus: false });
      index += 1;
      continue;
    }
    const character = value[index]!;
    out.push({ text: single[character] ?? character, nucleus: false });
  }

  // Вершина слога — гласный либо слоговое «r» между согласными («prst», «krv»).
  // Без второго правила такие слова выглядели бы вовсе без слогов.
  const isVowel = (sound: Sound | undefined) =>
    sound !== undefined && sound.text.length === 1 && IPA_VOWELS.includes(sound.text);

  for (let index = 0; index < out.length; index += 1) {
    const sound = out[index]!;
    if (isVowel(sound)) {
      sound.nucleus = true;
      continue;
    }
    if (sound.text !== 'r') continue;
    if (!isVowel(out[index - 1]) && !isVowel(out[index + 1])) sound.nucleus = true;
  }
  return out;
}

export function serbianIpa(word: string): string {
  const output = sounds(word)
    .map((sound) => sound.text)
    .join('');
  return output ? `/${output}/` : '';
}

/** Число слогов сербского слова. */
export function syllableCount(word: string): number {
  return sounds(word).filter((sound) => sound.nucleus).length;
}

// Знаки сербского ударения: краткое нисходящее (двойной гравис), долгое
// нисходящее (перевёрнутая бреве), краткое восходящее (гравис), долгое
// восходящее (акут). Макрон (U+0304) сюда не входит — он обозначает долготу
// безударного слога, и выделять по нему значит показать ударение не там.
const STRESS_MARKS = '̏̑̀́';

/**
 * Разрезает ударное написание словаря на части вокруг ударной буквы.
 *
 * «knjȉga» → knj / ȉ / ga. Ударный слог в сербском несёт диакритику, и это
 * единственное надёжное указание на место ударения: по буквам его не
 * восстановить.
 *
 * Проверяется, что знак стоит на гласной или на слоговом «r»: в разложенном
 * виде «ć» — это «c» плюс акут, то есть ровно тот же знак, что и долгое
 * восходящее ударение.
 */
export function splitAccented(written: string): IpaParts {
  const decomposed = written.normalize('NFD');
  let base = -1;
  for (let index = 0; index < decomposed.length; index += 1) {
    if (!STRESS_MARKS.includes(decomposed[index]!)) continue;
    const letter = decomposed[index - 1];
    if (!letter || !'aeiouraeiourАЕИОУРаеиоур'.includes(letter)) continue;
    base = index - 1;
    break;
  }
  if (base < 0) return { before: written, stressed: '', after: '' };

  let end = base + 1;
  while (end < decomposed.length && /\p{M}/u.test(decomposed[end]!)) end += 1;
  return {
    before: decomposed.slice(0, base).normalize('NFC'),
    stressed: decomposed.slice(base, end).normalize('NFC'),
    after: decomposed.slice(end).normalize('NFC'),
  };
}

/** Транскрипция, разрезанная на ударном звуке. */
export interface IpaParts {
  before: string;
  /** Ударный звук; пустая строка — место ударения неизвестно. */
  stressed: string;
  after: string;
}

/**
 * Транскрипция с отмеченным ударением.
 *
 * Место ударения в сербском по написанию не восстанавливается: ударений четыре
 * (краткое и долгое, восходящее и нисходящее), словаря ударений в проекте нет,
 * а поставить знак наугад значит учить неправильному произношению.
 *
 * Но в коротких словах место определяется однозначно, без всякого словаря:
 * ударение никогда не падает на последний слог (твёрдая норма литературного
 * сербского), поэтому в одно- и двусложном слове ударен первый слог. Там оно и
 * отмечается. В словах длиннее ударение может стоять на любом слоге, кроме
 * последнего, — там транскрипция остаётся без пометы: пустое место честнее
 * уверенной ошибки.
 */
export function serbianIpaParts(word: string): IpaParts {
  const all = sounds(word);
  if (all.length === 0) return { before: '', stressed: '', after: '' };

  const nuclei = all.flatMap((sound, index) => (sound.nucleus ? [index] : []));
  const at = nuclei.length > 0 && nuclei.length <= 2 ? nuclei[0]! : -1;
  const join = (from: number, to?: number) =>
    all.slice(from, to).map((sound) => sound.text).join('');

  if (at < 0) return { before: `/${join(0)}/`, stressed: '', after: '' };
  return {
    before: `/${join(0, at)}`,
    stressed: all[at]!.text,
    after: `${join(at + 1)}/`,
  };
}

File: frontend/lib/utils/serbian_pronunciation.dart
class SerbianPronunciation {
  const SerbianPronunciation._();

  static const _single = <String, String>{
    'a': 'a',
    'b': 'b',
    'c': 'ts',
    'č': 'tʃ',
    'ć': 'tɕ',
    'd': 'd',
    'đ': 'dʑ',
    'e': 'e',
    'f': 'f',
    'g': 'ɡ',
    'h': 'x',
    'i': 'i',
    'j': 'j',
    'k': 'k',
    'l': 'l',
    'm': 'm',
    'n': 'n',
    'o': 'o',
    'p': 'p',
    'r': 'r',
    's': 's',
    'š': 'ʃ',
    't': 't',
    'u': 'u',
    'v': 'ʋ',
    'z': 'z',
    'ž': 'ʒ',
    'а': 'a',
    'б': 'b',
    'в': 'ʋ',
    'г': 'ɡ',
    'д': 'd',
    'ђ': 'dʑ',
    'е': 'e',
    'ж': 'ʒ',
    'з': 'z',
    'и': 'i',
    'ј': 'j',
    'к': 'k',
    'л': 'l',
    'љ': 'ʎ',
    'м': 'm',
    'н': 'n',
    'њ': 'ɲ',
    'о': 'o',
    'п': 'p',
    'р': 'r',
    'с': 's',
    'т': 't',
    'ћ': 'tɕ',
    'у': 'u',
    'ф': 'f',
    'х': 'x',
    'ц': 'ts',
    'ч': 'tʃ',
    'џ': 'dʒ',
    'ш': 'ʃ',
  };

  static String ipa(String word) {
    final value = word.trim().toLowerCase();
    if (value.isEmpty) return '';
    final out = StringBuffer();
    for (var i = 0; i < value.length; i++) {
      if (i + 1 < value.length) {
        final pair = value.substring(i, i + 2);
        if (pair == 'dž') {
          out.write('dʒ');
          i++;
          continue;
        }
        if (pair == 'lj') {
          out.write('ʎ');
          i++;
          continue;
        }
        if (pair == 'nj') {
          out.write('ɲ');
          i++;
          continue;
        }
      }
      out.write(_single[value[i]] ?? value[i]);
    }
    return '/$out/';
  }

  static const _ipaVowels = 'aeiou';

  /// Звуки транскрипции: где вершины слогов, там и может стоять ударение.
  ///
  /// Слоговое «r» между согласными («prst», «krv») тоже вершина — без этого
  /// такие слова выходили бы вовсе без слогов.
  static List<(String, bool)> _sounds(String word) {
    final value = word.trim().toLowerCase();
    final out = <String>[];
    for (var i = 0; i < value.length; i++) {
      if (i + 1 < value.length) {
        final pair = value.substring(i, i + 2);
        const digraphs = {'dž': 'dʒ', 'lj': 'ʎ', 'nj': 'ɲ'};
        final mapped = digraphs[pair];
        if (mapped != null) {
          out.add(mapped);
          i++;
          continue;
        }
      }
      out.add(_single[value[i]] ?? value[i]);
    }

    bool isVowel(int index) {
      if (index < 0 || index >= out.length) return false;
      return out[index].length == 1 && _ipaVowels.contains(out[index]);
    }

    return [
      for (var i = 0; i < out.length; i++)
        (
          out[i],
          isVowel(i) || (out[i] == 'r' && !isVowel(i - 1) && !isVowel(i + 1)),
        ),
    ];
  }

  static int syllables(String word) =>
      _sounds(word).where((sound) => sound.$2).length;

  /// Транскрипция, разрезанная на ударном звуке: до, ударный, после.
  ///
  /// Место ударения в сербском по написанию не восстанавливается: ударений
  /// четыре (краткое и долгое, восходящее и нисходящее), словаря ударений в
  /// проекте нет, а поставить знак наугад значит учить неправильному
  /// произношению.
  ///
  /// Но в коротких словах место определяется однозначно и без словаря: ударение
  /// никогда не падает на последний слог — твёрдая норма литературного
  /// сербского, — поэтому в одно- и двусложном слове ударен первый. В словах
  /// длиннее ударный слог может быть любым, кроме последнего, и транскрипция
  /// остаётся без пометы: пустое место честнее уверенной ошибки.
  // Знаки сербского ударения. Макрон (U+0304) сюда не входит: он обозначает
  // долготу безударного слога, и выделять по нему значит показать ударение не
  // там.
  static const _stressMarks = '̏̑̀́';
  static const _accentable = 'aeiouraeiourАЕИОУРаеиоур';

  /// Разрезает ударное написание словаря на части вокруг ударной буквы.
  ///
  /// «knjȉga» → knj / ȉ / ga. Ударный слог несёт диакритику, и это единственное
  /// надёжное указание на место ударения — по буквам его не восстановить.
  ///
  /// Знак засчитывается только на гласной или слоговом «r»: в разложенном виде
  /// «ć» — это «c» плюс акут, то есть ровно тот же знак, что и долгое
  /// восходящее ударение.
  static (String, String, String) splitAccented(String written) {
    final units = written.runes.toList();
    var base = -1;
    for (var i = 1; i < units.length; i++) {
      if (!_stressMarks.contains(String.fromCharCode(units[i]))) continue;
      final letter = String.fromCharCode(units[i - 1]);
      if (!_accentable.contains(letter)) continue;
      base = i - 1;
      break;
    }
    if (base < 0) return (written, '', '');

    var end = base + 1;
    while (end < units.length &&
        RegExp(r'\p{M}', unicode: true)
            .hasMatch(String.fromCharCode(units[end]))) {
      end++;
    }
    String slice(int from, int to) =>
        String.fromCharCodes(units.sublist(from, to));
    return (slice(0, base), slice(base, end), slice(end, units.length));
  }

  static (String, String, String) ipaParts(String word) {
    final all = _sounds(word);
    if (all.isEmpty) return ('', '', '');

    final nuclei = [
      for (var i = 0; i < all.length; i++)
        if (all[i].$2) i,
    ];
    String join(int from, [int? to]) =>
        all.sublist(from, to ?? all.length).map((s) => s.$1).join();

    if (nuclei.isEmpty || nuclei.length > 2) return ('/${join(0)}/', '', '');
    final at = nuclei.first;
    return ('/${join(0, at)}', all[at].$1, '${join(at + 1)}/');
  }
}
